src/main/kotlin/minesweeper/domain/MineSweeper.kt

```diff
@@ -5,27 +5,67 @@ import minesweeper.domain.cell.MineCell
 import minesweeper.domain.cell.SafeCell
 
 class MineSweeper(
-    val mineSweeperSize: MineSweeperSize,
-    val minePosition: List<Int>
+    private val mineSweeperMap: Map<Position, Cell>
 ) {
-    val mineMap: Map<Int, List<Cell>> = (0 until mineSweeperSize.height).associateWith { y ->
-        (0 until mineSweeperSize.width).map { x ->
-            if (minePosition.contains(y * mineSweeperSize.height + x)) {
-                MineCell()
-            } else {
-                SafeCell(countAdjacentMine(x, y))
-            }
+    fun getRow(index: Int): List<Cell> {
+        require(index < getHeight()) { "Wrong index!" }
+        return mineSweeperMap.filter { (key, _) -> key.y == index }
+            .toSortedMap { prev, next -> prev.x.compareTo(next.x) }.values.toList()
+    }
+
+    fun openCell(position: Position): MineSweeperState {
+        val targetCell = getCell(position)
+
+        if (targetCell is MineCell) {
+            return MineSweeperState.LOSE
         }
+
+        openAdjacentCell(position)
+
+        if (countOfMine() == countOfClosed()) {
+            return MineSweeperState.WIN
+        }
+
+        return MineSweeperState.CONTINUE
     }
 
-    private fun countAdjacentMine(x: Int, y: Int): Int {
-        return Direction.eightWays.count { (dx, dy) ->
-            val newX = x + dx
-            val newY = y + dy
+    private fun countOfMine() = mineSweeperMap.values.count { it is MineCell }
+
+    private fun countOfClosed() = mineSweeperMap.values.count { !it.isOpened }
 
-            (0 until mineSweeperSize.width).contains(newX) &&
-                (0 until mineSweeperSize.height).contains(newY) &&
-                minePosition.contains(newX + newY * mineSweeperSize.height)
+    private fun openAdjacentCell(position: Position) {
+        val targetCell = getCell(position)
+        targetCell.open()
+
+        if (targetCell is SafeCell && targetCell.countOfAdjacentMine == 0) {
+            Direction.eightWays.map { (dx, dy) ->
+                Position(position.x + dx, position.y + dy)
+            }.forEach { it ->
+                if (isValidPosition(it) && !getCell(it).isOpened) {
+                    openAdjacentCell(it)
+                }
+            }
         }
     }
+
+    private fun isValidPosition(position: Position): Boolean {
+        return (0 until getWidth()).contains(position.x) &&
+            (0 until getHeight()).contains(position.y)
+    }
+
+    private fun getCell(position: Position): Cell {
+        val targetCell = mineSweeperMap[position]
+        require(targetCell != null) { "Wrong position!" }
+        return targetCell
+    }
+
+
+
+    private fun getWidth(): Int = mineSweeperMap.keys.count { (x, _) ->
```

Comment:
```suggestion
    private val width: Int 
        get() = mineSweeperMap.keys.count { (x, _) ->
```
getXXX는 프로퍼티로도 표현이 가능합니다. 실제로 이러한 프로퍼티들은 자바로 컴파일하면 getXXX와 같이 변환되구요.